packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -1096,12 +1097,31 @@ function updateContextProvider(
   return workInProgress.child;
 }
 
+let hasWarnedAboutUsingContextAsConsumer = false;
+
 function updateContextConsumer(
   current: Fiber | null,
   workInProgress: Fiber,
   renderExpirationTime: ExpirationTime,
 ) {
-  const context: ReactContext<any> = workInProgress.type;
+  let context: ReactContext<any> = workInProgress.type;
+  if (__DEV__) {
+    if (workInProgress.type._context === undefined) {
+      if (!hasWarnedAboutUsingContextAsConsumer) {
+        hasWarnedAboutUsingContextAsConsumer = true;
+        lowPriorityWarning(
+          false,
+          'You are using the Context from React.createContext() as a consumer.' +
+            'The correct way is to use Context.Consumer as the consumer instead. ' +
+            'This usage is deprecated and will be removed in the next major version.',
+        );
+      }
+    } else {
+      // Context.Consumer is a separate object in DEV, where
+      // _context points back to the original context
+      context = workInProgress.type._context;
```

Comment:
Can you explain why we read it differently in DEV and PROD? It would help to have a written explanation of DEV/PROD matrix for Consumer/Context object and how behavior changes for each.

Reply:
Can you reference me to another matrix? I updated the comment

Reply:
There's none. I just meant if you could write up of what's supposed to change.

* dev when using consumer: (before) and (after)
* dev when using context: (before) and (after)
* prod when using consumer: (before) and (after)
* prod when using context: (before) and (after)

That would be easier for me to review since I can focus on intent separately from code.

Reply:
I've put a big nice comment in there that explains things better now. I'll aadd the matrix to this PR.